### PR TITLE
Unwrapping Use-site Targets in Kotlin

### PR DESCRIPTION
**Pull request by Shpota**

### CODE REVIEW

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+title:  "Unwrapping Use-site Targets in Kotlin"
+date:   2021-03-02 10:00:00 +0200
+comments: true
+img:
+  href: 2021-03-02-dogs-of-berlin.png
+  copyright: Dogs of Berlin by Christian Alvart
+  alt: Dogs of Berlin
+---
+I like Kotlin for its expressiveness. Its interoperability with Java
+allows to easily use the existing Java ecosystem while enjoying a modern
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:12]: "allows for an easy use of..."

Comment (rebeccaandersonediting) [inline _posts/2021-02-20-kotlin-use-site-targets.md:12]: "allows **the user** to easily use the existing Java ecosystem ...

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+language. When it comes to annotations Kotlin handles them almost the same
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:13]: "When it comes to annotations, Kotlin..." 

Added a comma after "annotations"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+as Java. In Kotlin, a single syntax construction might be compiled into
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:14]: "Kotlin handles them in almost the same way as Java does."

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+several JVM constructions. A Kotlin property would produce a JVM field,
+a getter, and optionally a setter. This is where annotation use-site targets
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:16]: add a comma after "optionally"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    @Unique
+    val name: String,
+    @Size(min = 18, max = 100)
+    val age: Int,
+)
+```
+
+Once it is compiled, the corresponding class would have a constructor,
+fields, getters, equals, hashcode, and some other language elements.
+If you need that in the compiled code annotations were generated specifically
```
Comment (rebeccaandersonediting) [inline _posts/2021-02-20-kotlin-use-site-targets.md:34]: If you need that in the compiled code, annotations were generated specifically...

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+for getters or fields you can specify so-called use-site targets:
```
Comment (rebeccaandersonediting) [inline _posts/2021-02-20-kotlin-use-site-targets.md:35]: ...fields. You can specify so-called use-site targets:

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+use-site target an annotation has to use. In fact, forgetting to specify an
+annotation target is one of the most common sources of bugs in my code.
+
+It would be great if I could omit to specify use-site targets, and
+the Kotlin compiler would magically put them into the proper places.
+
+This is exactly the reason why I decided to create 
+[Ktargeter](https://github.com/ktargeter/ktargeter). I spent
+some time researching the subject. I tried annotation processing,
+code generation, but eventually, I went with creating a compiler
```
Comment (rebeccaandersonediting) [inline _posts/2021-02-20-kotlin-use-site-targets.md:59]: I tried annotation processing and

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:59]: code generation, but eventually I went with ...

File: _posts/2021-02-20-kotlin-use-site-targets.md (hunk 6 above)
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:35]: "If you need that in the compiled code, where annotations are generated specifically for getters or fields, you can specify the so-called use-site targets"
OR
"If you needed that in the compiled code, annotations were generated specifically for getters or fields, you could specify the so-called use-site targets"

Use either one of the corrected sentences based on what you mean.

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+```kotlin
+data class User(
+    @field:Unique
+    val name: String,
+    @get:Size(min = 18, max = 100)
+    val age: Int,
+)
+```
+In the compiled code the `name` field will be annotated with `@Unique`
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:45]: Add a comma after "code"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+and a getter of the `age` property will have the `@Size` annotation.
+
+This all works great unless you forget to put the corresponding target in front
+of an annotation. The problem is that there is no intuitive way to find out what
+use-site target an annotation has to use. In fact, forgetting to specify an
+annotation target is one of the most common sources of bugs in my code.
+
+It would be great if I could omit to specify use-site targets, and
```
Comment (rebeccaandersonediting) [inline _posts/2021-02-20-kotlin-use-site-targets.md:53]: It would be much more efficient if I could omit the use-site target specifications and the Kotlin compiler magically put them into the proper places.

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    @field:Unique
+    val name: String,
+    @get:Size(min = 18, max = 100)
+    val age: Int,
+)
+```
+In the compiled code the `name` field will be annotated with `@Unique`
+and a getter of the `age` property will have the `@Size` annotation.
+
+This all works great unless you forget to put the corresponding target in front
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:48]: "All this works great, as long as you don't forget to put the corresponding target in front of an annotation.

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+of an annotation. The problem is that there is no intuitive way to find out what
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:49]: Add a comma after "that"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+use-site target an annotation has to use. In fact, forgetting to specify an
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:50]: Add a comma after "use-site target"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+annotation target is one of the most common sources of bugs in my code.
+
+It would be great if I could omit to specify use-site targets, and
+the Kotlin compiler would magically put them into the proper places.
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:54]: "...would magically/automatically put them in proper places."

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+This is exactly the reason why I decided to create
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:56]: "This is exactly why I decided to create"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+of an annotation. The problem is that there is no intuitive way to find out what
+use-site target an annotation has to use. In fact, forgetting to specify an
+annotation target is one of the most common sources of bugs in my code.
+
+It would be great if I could omit to specify use-site targets, and
+the Kotlin compiler would magically put them into the proper places.
+
+This is exactly the reason why I decided to create 
+[Ktargeter](https://github.com/ktargeter/ktargeter). I spent
+some time researching the subject. I tried annotation processing,
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:58]: "...researching on the subject."

File: _posts/2021-02-20-kotlin-use-site-targets.md (hunk 7 above)
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:59]: "...and code generation, but I eventually went with creating a compiler..."

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+code generation, but eventually, I went with creating a compiler
+plugin. Luckily, Kotlin 1.4 brought [the new JVM IR Backend](https://blog.jetbrains.com/kotlin/2021/02/the-jvm-backend-is-in-beta-let-s-make-it-stable-together/)
+which is a set of APIs that allow to easily preprocess Kotlin code.
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:61]: Use "enable" instead of "allow"

File: _posts/2021-02-20-kotlin-use-site-targets.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+Long story short, if you don't want to specify annotation use-site
+targets or you keep forgetting to use them, you can configure it once
```
Comment (SudarshanBK) [inline _posts/2021-02-20-kotlin-use-site-targets.md:64]: "targets or **if** you keep forgetting..."

File: _posts/2021-02-20-kotlin-use-site-targets.md (hunk 6 above)
Comment (Shpota) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: @rebeccaandersonediting I guess it changes the meaning of the sentence. The "you can specify..." part is related to the previous statement  staring with "If you need". 

It is like "If you need ... **then** you can specify...". Could you please rephrase it to make clear?

**Review by SudarshanBK — COMMENTED** [review]
Hey, I am Sudarshan from UpWork. I had submitted a proposal to work on editing your blog, but went a step ahead and finished the work for you too. I have suggested changes in some of the sentences and adding punctuations in some places, as requested. Message me on UpWork and let me know if this worked! Hope you accept my proposal.

File: _posts/2021-02-20-kotlin-use-site-targets.md (hunk 1 above)
Comment (Shpota) [reply _posts/2021-02-20-kotlin-use-site-targets.md:12]: It is more like a developer, not a user. If I replaced "the user" with "us" would it sound better?

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:12]: Us is tricky with informal texts. Could you say, "allows the developer to easily use the existing Java ecosystem..."

File: _posts/2021-02-20-kotlin-use-site-targets.md (hunk 6 above)
Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: "If you need that in compiled code, you can specify so-called use-site targets. Annotations were generated specifically for getters or fields to allow for this."  OR "If you need that in compiled code, annotations generated specifically for getters or fields allow you to specify so-called use-site targets."

Comment (Shpota) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: @rebeccaandersonediting these options also change the meaning. I am not sure if I am able to explain it properly without much context, but let me try. If a user needs "to generate annotations specifically for getters or fields" then the user can "specify so-called use-site targets". In other words, specifying use-site targets will allow the user to have annotations generated specifically for getters or setters in the compiled code.

Does it make sense?

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: If you need that in compiled code, you must generate annotations specifically for getters or fields, which will then allow you to specify so-called use-site targets.

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: "If you need that in the compiled code, annotations must be generated specifically for getters or fields in order to allow you to specify so-called use-site targets"

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: By generating annotations specifically for getters or fields, you will be able to include that in the compiled code and specify so-called use-site targets.

Comment (rebeccaandersonediting) [reply _posts/2021-02-20-kotlin-use-site-targets.md:35]: "If you need that in the compiled code, you must specify the so-called use-site targets in order to generate annotations specifically for getters or fields."
